Add cache support for SFMC oauth tokens

pmac opened the pull request.

pmac commented:
@jgmize something like this work you think?

jgmize commented on news/backends/sfmc.py:
```diff
@@ -11,19 +12,54 @@
+        for prop in self.token_property_names:
+            new_val = getattr(self, prop)
+            if new_val is not None and new_val != getattr(self, '_old_' + prop):
+                cache.set(self._cache_key(prop), new_val, 3600)
```
What is the advantage of iterating through all of the token_property_names and storing them under individual cache keys vs storing all of them in a dict with a single cache key?

pmac replied:
Hmm... dunno. I'll try that.

pmac commented:
Okay. I've changed the strategy to cache a dict of all of the token info as well as further improved the `refresh_token` method (copied it from the base class and modified it as it was just not great code to start with, and this is much cleaner, and I get to use the full response from the API). I also added a bit of Thundering Herd protection to the general cache as well as the token refresh.

jgmize commented on news/backends/sfmc.py:
```diff
@@ -10,19 +12,95 @@
+        tokens = cache.get(self.token_cache_key)
+        if tokens:
+            for prop, value in tokens.items():
+                setattr(self, prop, value)
```
It might be worth adding a little validation here, like `if prop in self.token_property_names`

pmac replied:
there should be no way for an item outside of `token_property_names` to get into the cached dict. Are you thinking we should treat this data as potentially hostile?

jgmize replied:
not necessarily hostile, but we're crossing a network here and bit flips and other data corruption are possible

jgmize commented on news/backends/sfmc.py:
```diff
@@ -10,19 +12,95 @@
+            if self.refreshKey:
+                payload['refreshToken'] = self.refreshKey
+
+            while True:
```
NIT: this works, but to me it would be more readable if we refactored this while loop into a recursive function, something like:

```python
def request_token(self, payload):
    r = requests.post(self.auth_url, json=payload)
    token_response = r.json()
    if 'accessToken' in token_response:
        return token_response
    elif 'refreshToken' in payload:
        del payload['refreshToken']
        return request_token(payload)
    else:
        raise NewsletterException(
            'Unable to validate App Keys(ClientID/ClientSecret) provided: ' +
            repr(token_response), status_code=r.status_code)
```

pmac commented:
I dig the recursive function idea. I also felt a little dirty with the infinite loop. It looks good I think.

pmac commented:
Added some tests. I think we're good. Want a final look @jgmize ?

jgmize commented:
r+, nice work @pmac
